### PR TITLE
Hacking on Atom Core: Update build requirements to be accurate

### PR DESCRIPTION
**Pull request by DeeDeeG**
### Requirements for Contributing Documentation

* Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* The pull request must only contribute documentation (for example, markdown files or API docs). To contribute other changes, you must use a different template. You can see all templates at <https://github.com/atom/.github/tree/master/.github/PULL_REQUEST_TEMPLATE>.

### Description of the Change

Information on required Visual Studio versions and Python versions were out of date, considering that we use newer `node-gyp` now.

See: https://github.com/nodejs/node-gyp/tree/v5.x/#installation for a more up-to-date reference on the required Visual Studio and Python versions, and other miscellaneous requirements of `node-gyp`.

I updated "Hacking on Atom Core" to reflect that **Python 3** now works (See: https://github.com/atom/atom/pull/20711). **Visual Studio 2017** has also been supported by recent `node-gyp` for quite some time. Relatedly, on Windows 10 the **Windows 10 SDK** is available, not the **Windows 8 SDK**. **Node v10.12+** is required, and NPM < 6 is no longer supported by the NPM authors, so I updated the documentation to reflect that as well.

### Release Notes

N/A (Atom Flight Manual doesn't appear to compile or publish any release notes)

### CODE REVIEW

**Comment by DeeDeeG** [comment]
I was tempted to update the Node/NPM versions to supported versions (Node 10 or newer, NPM 6 or newer). I guess the older versions might work (it suggests Node 6 and NPM 3, which are very old by now and no-longer supported) but I haven't tested that yet.

(The norm in the Node ecosystem is to point people toward supported versions of Node, and to a lesser extent to point folks toward supported versions of modules such as `npm`. I think it would be resonable to update those to v10+ and v6+ respectively, and I'd be happy to do so in this PR if a reviewer here was alright with that.)

File: content/hacking-atom/sections/hacking-on-atom-core.md [hunk 1]
```diff
@@ -85,31 +85,30 @@ In order to build Atom from source, you need to have a number of other requireme
...
 * Command Line Tools for [Xcode](https://developer.apple.com/xcode/downloads/) (run `xcode-select --install` to install)
 
 {{/mac}}
 
 {{#windows}}
 
 * Node.js 6.9.4 or later (the architecture of node available to the build system will determine whether you build 32-bit or 64-bit Atom)
-* Python v2.7.x
+* Python v2.7.x or 3.5+
   * The `python.exe` must be available at `%SystemDrive%\Python27\python.exe`. If it is installed elsewhere create a symbolic link to the directory containing the `python.exe` using: `mklink /d %SystemDrive%\Python27 D:\elsewhere\Python27`
```
Comment (DeeDeeG) [inline content/hacking-atom/sections/hacking-on-atom-core.md:97]: This part is sadly still true on Windows, until https://github.com/atom/atom/pull/20856 is merged.

The bootstrap script errors out if Python 2.7 isn't found at `[Drive_root]:\Python27\python.exe`. This is hard-coded in Atom's `script/lib/verify-machine-requirements.js` sub-script. My PR that I just linked to above fixes that.

**Comment by DeeDeeG** [comment]
I found that Atom requires Node v10.12+ to build now. This comes with npm 6 bundled with it. So that should probably be updated as well.

---

Research:

Atom has an indirect dependency `extract-zip@^2`, which requires Node v10.12+ due to using recursive `mkdir`.

- https://github.com/nodejs/node/blob/v10.x/doc/changelogs/CHANGELOG_V10.md#10.12.0
  - Node v10.12 adds a recursive option for `fs.mkdir`
- `extract-zip@2` uses recursive `fs.mkdir`, thus requiring Node v10.12+

Atom indirectly requires `extract-zip@2` as of this PR: https://github.com/atom/atom/pull/20879
- Atom requires `electron-chromedriver@^9.0.0` and `electron-mksnapshot@^9.0.2`, which both depend on `extract-zip@2`.

**Comment by DeeDeeG** [comment]
One last thing to note, Python 3 only works with `npm` v6.12+ https://github.com/npm/cli/blob/latest/CHANGELOG.md#6120-2019-10-08, which was the first version of `npm` with `node-gyp` v5.0.5+ https://github.com/nodejs/node-gyp/blob/master/CHANGELOG.md#v505-2019-10-04 which is in turn the earliest version of `node-gyp` that works with Python 3 "out of the box" (no experimental flags required).

So maybe that `npm` version requirement for using Python 3 should be explained here in the docs, or to keep it simple, just say Atom requires `npm` 6.12+.

**Comment by DeeDeeG** [comment]
Travis test failure seems unrelated. It apparently timed out trying to visit https://bugs.chromium.org/p/gyp/issues/detail?id=393, which I can visit just fine. Just a temporary network issue I guess.

---

Somewhat tangential/off-topic: That old bug has most of its content deleted. I wonder if it's been fixed? I can't tell if the upstream issue at `gyp` ever got fixed, reading the related `node-gyp` issue: https://github.com/nodejs/node-gyp/issues/297 So maybe the part of "Hacking on Atom Core" about that error can be deleted.

File: content/hacking-atom/sections/hacking-on-atom-core.md [hunk 2]
```diff
@@ -83,33 +83,32 @@ In order to build Atom from source, you need to have a number of other requireme
...
 * 7zip (7z.exe available from the command line) - for creating distribution zip files
 * C++ build tools, either:
-  * [Visual C++ Build Tools 2015](https://visualstudio.microsoft.com/vs/older-downloads/)
-  * [Visual Studio 2013 Update 5](https://www.visualstudio.com/downloads/) (Express Edition or better)
-  * [Visual Studio 2015](https://www.visualstudio.com/downloads/) (Community Edition or better)
+  * [Visual C++ Build Tools 2015 or newer](https://visualstudio.microsoft.com/visual-cpp-build-tools/)
+  * [Visual Studio 2015 or newer](https://www.visualstudio.com/downloads/) (Community Edition or better)
   * **Unsupported** but more convenient for some who know Node: [windows-build-tools](https://www.npmjs.com/package/windows-build-tools)
```
Comment (aminya) [inline content/hacking-atom/sections/hacking-on-atom-core.md:102]:
```suggestion
  * [windows-build-tools](https://www.npmjs.com/package/windows-build-tools)
```
This is what we use in the CI builds and we have not had any issues with it. Why we should consider this as unsupported?

https://github.com/atom/atom/blob/master/script/vsts/platforms/windows.yml#L42-L44

File: content/hacking-atom/sections/hacking-on-atom-core.md [hunk 3]
```diff
@@ -83,33 +83,32 @@ In order to build Atom from source, you need to have a number of other requireme
...
+* Python v2.7.x or 3.5+
 * Command Line Tools for [Xcode](https://developer.apple.com/xcode/downloads/) (run `xcode-select --install` to install)
 
 {{/mac}}
 
 {{#windows}}
 
-* Node.js 6.9.4 or later (the architecture of node available to the build system will determine whether you build 32-bit or 64-bit Atom)
-* Python v2.7.x
+* Node.js 10.12 or later (the architecture of node available to the build system will determine whether you build 32-bit or 64-bit Atom)
```
Comment (aminya) [inline content/hacking-atom/sections/hacking-on-atom-core.md:95]: We use Node 12 in our CI builds. Shouldn't that be the supported version?

Comment (DeeDeeG) [reply content/hacking-atom/sections/hacking-on-atom-core.md:95]: 10 is still an officially supported Long Term Support version for now: https://nodejs.org/about/releases/

(Until the end of April in 2021)

File: content/hacking-atom/sections/hacking-on-atom-core.md (hunk 2 above)
Comment (DeeDeeG) [reply content/hacking-atom/sections/hacking-on-atom-core.md:102]: I agree. This method is so much easier, and as you point out, it's literally the way Atom is tested in CI.

If it's good enough for CI, it should be good enough for end-users. It should be the first option suggested, in my opinion. Would be interested in feedback from the Atom maintainers.

Comment (aminya) [reply content/hacking-atom/sections/hacking-on-atom-core.md:102]: That's how the official releases are built! 😄 I think this should even be written as the first option.

**Comment by DeeDeeG** [comment]
@darangi, @lkashef:

we (@aminya and I) are trying to work on Atom and polish up all the rough edges for those who want to fork it. This important documentation explaining how to get started with Atom is one of the first things users read when trying to contribute. We would really appreciate if these updates could be reviewed and approved.

These suggested updates are based on our active contributions to Atom, including my work on updating some of Atom's build script Node dependencies, and both of our reading through/running the CI configs. I know them to be accurate. It would be great to get updated docs.

Thanks for considering.

\- DeeDeeG

**Comment by darangi** [comment]
@DeeDeeG and @aminya Thanks for the contribution 🙇‍♂️ 

We can't provide an accurate timeline but we will definitely put this on our radar and act on it as soon as possible.

**Comment by DeeDeeG** [comment]
More research about which NPM versions are supported:

Briefly: Users with **_NPM 3.10.7 or newer_** can bootstrap and build Atom successfully.

(I personally still lean toward recommending folks should use the latest major version of NPM, that is currently **_NPM 6_** (or newer), which is what I did in this PR, but I am open to feedback from the maintainers.)

<details><summary>Research details (click to expand):</summary>

Tested on Ubuntu 20.04: Atom can be bootstrapped and built with older NPM all the way back to `npm@3.10.7`.

This is a little alarming to do, in my opinion, because [`package-lock.json` was introduced in `npm@5`](https://github.com/npm/npm/releases/tag/v5.0.0). So the `script/node_modules` dependencies are installed with an older dependency tree algorithm without the benefit of the lockfile.

Somehow this still works, at least for bootstrap/build. This doesn't affect the `apm/node_modules` install or the main/App dependencies in `[repo_root]/node_modules`. The first step installs `script/` dependencies (including `npm@6`) with `npm@3.10.7`. Then the `script/node_modules/.bin/npm` (`npm@6` ) is used to install `apm`. Then `apm` (with its built-in Node@10 and NPM@6) installs the main Atom dependencies, including bundled packages such as `atom-dark-ui`.

So overall the user's NPM isn't used for much, and apparently only needs to be v3.10.7 or greater.

By the way, newer versions of NPM are supposedly much faster to complete an `npm install`, so users should want to benefit from that by using NPM 5 or newer.

And besides: Node 10.12.0, the oldest Node we support right now, already comes bundled with NPM 6.4.1... It's hard to picture someone running Node 10.12 or newer with an NPM older than that.

I'm thinking we should still recommend users use NPM@6 or greater. Older versions aren't supported by the NPM team, for one reason. And for another reason, anything older than NPM@5 won't read the `package-lock.json` lockfile in `script/package-lock.json`.

</details>

**Comment by DeeDeeG** [comment]
Research about Visual Studio:

node-gyp 5.x supports Visual Studio 2015, 2017 and 2019, at least officially. (See the [CHANGELOG.md](https://github.com/nodejs/node-gyp/blob/v5.x/CHANGELOG.md#v500-2019-06-13). Visual Studio 2019 support was added in node-gyp 5.0.0.)

---

With the versions of Node we are using, Visual Studio 2013 is no longer supported. If you try to do `node-gyp configure --verbose` on Node >= 8, you see this message in the log:

`not looking for VS2013 as it is only supported up to Node.js 8`

**Comment by aminya** [comment]
I cannot build Atom using 2019 preview. I believe that's the same as 2019.

File: content/hacking-atom/sections/hacking-on-atom-core.md [hunk 4]
```diff
@@ -83,33 +83,32 @@ In order to build Atom from source, you need to have a number of other requireme
...
 
   Also ensure that:
 
   * The default installation folder is chosen so the build tools can find it
   * If using Visual Studio make sure Visual C++ support is selected/installed
-  * If using Visual C++ Build Tools make sure Windows 8 SDK is selected/installed
+  * If using Visual C++ Build Tools make sure a Windows SDK (Windows 8 SDK or Windows 10 SDK) is selected/installed
   * A `git` command is in your path
-  * Set the `GYP_MSVS_VERSION` environment variable to the Visual Studio/Build Tools version (`2013` or `2015`) e.g. ``[Environment]::SetEnvironmentVariable("GYP_MSVS_VERSION", "2015", "User")`` in PowerShell (or set it in Windows advanced system settings).
+  * Set the `GYP_MSVS_VERSION` environment variable to the Visual Studio/Build Tools version (`2015`, `2017`, or `2019`.) e.g. ``[Environment]::SetEnvironmentVariable("GYP_MSVS_VERSION", "2015", "User")`` in PowerShell (or set it in Windows advanced system settings).
```
Comment (DeeDeeG) [inline content/hacking-atom/sections/hacking-on-atom-core.md:111]: Setting `GYP_MSVS_VERSION` is usually unnecessary, if you've installed to the default location.

node-gyp has [gotten better at locating Visual Studio](https://github.com/nodejs/node-gyp/pull/1762) and its components.

Maybe leave this for the troubleshooting area instead. It's already mentioned there.

**Comment by DeeDeeG** [comment]
> I cannot build Atom using 2019 preview. I believe that's the same as 2019.

I just built Atom with Visual Studio Build Tools 2019 to make sure it works.

The "About" screen of Atom 1.50.0-dev-110b05baf x64, on Windows
Visual Studio Build Tools 2019 Installer version 16.6.3, with VS 2019 C++ build tools and Windows 10 SDK components selected

When I do [More] > [Export configuration], I get this:

```yaml
{
  "version": "1.0",
  "components": [
    "Microsoft.VisualStudio.Component.Roslyn.Compiler",
    "Microsoft.Component.MSBuild",
    "Microsoft.VisualStudio.Component.CoreBuildTools",
    "Microsoft.VisualStudio.Workload.MSBuildTools",
    "Microsoft.VisualStudio.Component.VC.Tools.x86.x64",
    "Microsoft.VisualStudio.Component.Windows10SDK.18362"
  ]
}
```

File: content/hacking-atom/sections/hacking-on-atom-core.md [hunk 5]
```diff
@@ -83,33 +83,36 @@ In order to build Atom from source, you need to have a number of other requireme
...
+* Node.js 10.12 or later (the architecture of node available to the build system will determine whether you build 32-bit or 64-bit Atom)
+* Python v2.6.x, v2.7.x, or v3.5+
+  * [Get Python from the Microsoft Store](https://www.microsoft.com/en-us/search/shop/apps?q=python+software+foundation&devicetype=pc&category=Developer+tools%5cDevelopment+kits&Price=0&MaturityRating=ESRB%3aE), or
+  * Download Python from https://www.python.org/downloads/.
+    * For Python 2, be sure to install in the default location, or check "Add Python 2.x to PATH" before installing.
+    * For Python 3, check "Add Python 3.x to PATH", or change the install path to `[Your_Drive_Letter]:\Python37` e.g. `C:\Python37`, (even if your version of Python 3 isn't 3.7, that's one place where the scripts will look.) 
+    * If python isn't found by the bootstrap script, create a symbolic link to the directory containing `python.exe` using e.g.: `mklink /d %SystemDrive%\Python27 D:\elsewhere\Python27`(Links should be set at either `%SystemDrive%\Python27` or `%SystemDrive%\Python37`, regardless of what version of Python you actually have.)
```
Comment (DeeDeeG) [inline content/hacking-atom/sections/hacking-on-atom-core.md:101]: I'm not sure if this is too much detail. It should be accurate, though.

File: content/hacking-atom/sections/hacking-on-atom-core.md [hunk 6]
```diff
@@ -83,33 +83,36 @@ In order to build Atom from source, you need to have a number of other requireme
...
 * 7zip (7z.exe available from the command line) - for creating distribution zip files
```
Comment (DeeDeeG) [inline content/hacking-atom/sections/hacking-on-atom-core.md:102]: 7-zip is a dependency in `scripts/package.json` now, as of https://github.com/atom/atom/pull/17538.

7-Zip as a system installed program wouldn't be used. So it doesn't need to be installed anymore before bootstrapping/building on Windows.

- https://github.com/atom/atom/pull/17538/files#diff-0f080e84d61cfb6f244618e43d478535R5
- https://github.com/atom/atom/pull/17538/files#diff-a2da2402b4e11152e22400e2eb7f52e8R6
- https://github.com/atom/atom/pull/17538/files#diff-a2da2402b4e11152e22400e2eb7f52e8L47-R48

File: content/hacking-atom/sections/hacking-on-atom-core.md (hunk 1 above)
Comment (DeeDeeG) [reply content/hacking-atom/sections/hacking-on-atom-core.md]: atom/atom#20856 was merged! 🎉

File: content/hacking-atom/sections/hacking-on-atom-core.md (hunk 2 above)
Comment (DeeDeeG) [reply content/hacking-atom/sections/hacking-on-atom-core.md]: It's the first recommendation straight from a Microsoft guide document as well: https://github.com/Microsoft/nodejs-guidelines/blob/master/windows-environment.md#environment-setup-and-configuration

The document refers to it as "Microsoft's windows-build-tools":

> Option 1: Install all the required tools and configurations using Microsoft's [windows-build-tools] by running `npm install -g windows-build-tools` from an elevated PowerShell (run as Administrator).

**Comment by sadick254** [comment]
@DeeDeeG Thank you for your submission. :tada:

Have you considered all the possible combinations and ensured that all the them work?
It would be really helpful for us to have a list of the combinations you have tried and the results.

**Comment by DeeDeeG** [comment]
@sadick254 yeah, I can do some testing on this.

There's definitely a lot of "moving parts" so to speak, considering these:
- Python version
- Visual Studio version
- Node version
- npm version

<details><Summary>Research details for which versions of npm, Node, Python and Visual Studio are supported (click to expand):</summary>

### npm:

Thankfully, the system `npm` is _only_ used to install the build scripts. The build scripts download a copy of exactly `npm 6.14.4`. Then that is used to install `apm`, which also includes an version of npm in the `6.14.x` series. ([6.14 is the latest `npm`](https://github.com/npm/cli/releases) at the moment, so that's all good.)

**After the `script/node_modules` dependencies are installed, we are running modern `npm` and modern `node-gyp`.**

Through testing, I have [found that npm 3.10.7 is the oldest `npm` that can install the build scripts](https://github.com/atom/flight-manual.atom.io/pull/630#issuecomment-656279220), after which the build scripts' own npm 6.14.4 is used.

_(Note: Version 3.10 is a really old version of `npm`, so telling users to use something newer might be a good idea anyway. Only the latest release of `npm` receives any support or fixes, as far as I can tell.)_

#### in short: `npm` 3.10.7+ can be used.

### Node:

The role of system Node in the bootstrap/build process is to: Run the build scripts and run `npm`, up to and including installing the `scripts/` dependencies, and then installing the `apm/` dependencies.

I have found that Node 10.12.0 or newer is required, because one of our `script/` dependencies uses [recursive `fs.mkdir`](https://nodejs.org/api/fs.html#fs_fs_mkdir_path_options_callback), which was [introduced in Node 10.12.0](https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V10.md#10.12.0).

_(Note: As far as I can tell, any version of Node can successfully run any version of `npm`. But the test matrix to test every combination of Node 10.12.0+ and `npm` 3.10.7+ would be enormous, so sadly I cannot commit to investigating every combination of this.)_

**In short, Node 10.12.0+ can be used.**

### Python:

This one is easy for me to confirm.

`node-gyp` sets our Python requirements. And we get `node-gyp` as a dependency of `npm`.

<details><summary>Info and background about node-gyp (click to expand):</summary>

`npm` comes with a native build system for C++ code, which is `node-gyp`. `node-gyp` uses [Google's `gyp` ("Generate Your Projects")](https://en.wikipedia.org/wiki/GYP_(software)) build system, which is written in python. So, `node-gyp` needs a python interpreter on the system in order to run. `node-gyp` sets our Python requirements. (_Note: Almost nowhere else in the Node_)

</details>

The first version of `node-gyp` to detect and use Python 3 (without having to set some "experimental Python 3" flag at runtime) is [`node-gyp` 5.0.5](https://github.com/nodejs/node-gyp/blob/master/CHANGELOG.md#v505-2019-10-04), as of this PR: https://github.com/nodejs/node-gyp/pull/1910.

They don't allow Python 3 older than 3.5, or older than 2.6... _(Note: In upcoming versions of node-gyp, Python 2.6 is dropped and only Python 2.7 is supported in the Python 2 series. [This is tentatively expected for `npm` 7.0](https://github.com/npm/cli/issues/1281))._

`npm` 6.12.0 is the first version to include new enough `node-gyp` (`node-gyp` 5.0.5) to detect and support Python 3 out of the box. See the Changelog to confirm: https://github.com/npm/cli/blob/latest/CHANGELOG.md#6120-2019-10-08 (look under "DEPENDENCIES" subheading).

Research findings:

- The oldest `npm` that Atom developers can use is `npm` 3.10.7, see above. [In npm 3.10.7, node-gyp 3.4 is included](https://github.com/npm/cli/blob/v3.10.7/package.json#L65). [In node-gyp 3.4, Python 2.5+ (but not Python3) can be used](https://github.com/nodejs/node-gyp/blob/v3.4.0/lib/configure.js#L433)
- Developers with `npm` 6.12.0+ can use Python 3.5+ or Python 2.6+.
- Developers with modern `npm` less than 6.12 can use Python 2.6+ _(no Python 3)_
- Developers with `npm` 6.12.0+ can use Python 3.5+ or Python 2.6+
- In the future, when `npm` 7 is released, Python 2.6 support may be dropped. Developers with `npm` 7 will need Python 3.5+ or 2.**7**+

**In short: We can say Python 2.7 is supported, and that Python 3.5+ is supported if you have `npm` 6.12.0 or newer.**

### Visual Studio version:

Again, this is tied to node-gyp.

Timeline:

- Oldest `node-gyp` we can use is 3.4.0 (from [`npm` 3.10.7](https://github.com/npm/cli/blob/v3.10.7/package.json#L65))
  - Node-gyp 3.4 appears to support Visual Studio 2005 - 2015
  - https://github.com/nodejs/node-gyp/blob/v3.4.0/gyp/pylib/gyp/MSVSVersion.py#L334-L350
- Node-gyp 3.6 (from [`npm` 4.5.0](https://github.com/npm/cli/blob/latest/changelogs/CHANGELOG-4.md#dependency-updates-1)) can find Visual Studio 2017
  - https://github.com/nodejs/node-gyp/blob/master/CHANGELOG.md#v360-2017-03-16
  - This PR: https://github.com/nodejs/node-gyp/pull/1130
- Node-gyp 5.x is used in all modern `npm` versions.
  - Node-gyp 5.0.0 (from [`npm` 6.10.1](https://github.com/npm/cli/blob/latest/CHANGELOG.md#v6101-2019-07-11)) added support for Visual Studio 2019
  - Also greatly improved detection for all versions of Visual Studio 
  - https://github.com/nodejs/node-gyp/blob/master/CHANGELOG.md#v500-2019-06-13
  - In this PR: https://github.com/nodejs/node-gyp/pull/1762
    - Files in this PR do not allow Using Visual Studio 2013 or older on Node greater than v8.
      - https://github.com/nodejs/node-gyp/blob/968c9067d7a25a80925b9f7a342b67a04c3f76d8/lib/find-visualstudio.js#L348-L352
    - The files from this PR also don't attempt to look for anything older than Visual Studio 2013, so the minimum VS version is now VS 2013
      - https://github.com/nodejs/node-gyp/blob/968c9067d7a25a80925b9f7a342b67a04c3f76d8/lib/find-visualstudio.js#L62-L77
- Node-gyp 5.0.5 (from [`npm` 6.12.0](https://github.com/npm/cli/blob/latest/CHANGELOG.md#6120-2019-10-08)) added support for a niche Visual Studio version, Visual Studio 2017 express
  - https://github.com/nodejs/node-gyp/blob/master/CHANGELOG.md#v505-2019-10-04
  - In this PR: https://github.com/nodejs/node-gyp/pull/1902

_(Note: `npm` bumps their `node-gyp` dependency somewhat often, so pretty much all versions of `node-gyp` were picked up by `npm` at some point.)_

**In short: We can say Visual Studio 2015 is supported. And Visual Studio 2017 is supported if you have npm 4.5 or newer. And Visual Studio 2019 is supported if you have npm 6.10.1 or newer.**

</details>

### Conclusion:

**According to researching the documentation, these versions can be used:** Node 10.12+, npm 3.10.7+, Python 2.6/2.7 (Python 3.5+ can be used with recent npm), Visual Studio 2015 (Visual Studio 2017 and 2019 can be used with recent npm).

I can test some combinations within those ranges and let you know how it goes.

**Comment by DeeDeeG** [comment]
> Visual Studio 2019 does not fully work as mentioned in:
> atom-ide-community/atom#117

I suppose we can hold back on recommending Visual Studio 2019 for now.

(I don't fully know whether atom-ide-community/atom#117 is a CI-only issue. I could try to get passing tests locally after building with VisualStudio Build Tools 2019, but it takes quite a long time to run all the tests, and there are timeouts and intermittent failures... I honestly find it hard to ever get a clean passing test run under any circumstances. So I may skip verifying that Visual Studio 2019 works 100% and just not advertise Visual Studio 2019 compatibility. Eventually we will need to cross that bridge, but Visual Studio 2015 is supported into the year 2025, Visual Studio 2017 is supported into the year 2027.

https://docs.microsoft.com/en-us/visualstudio/releases/2019/servicing#support-for-older-versions-of-visual-studio)

> The Node version should be fixed to 12.4.0 to prevent the issues mentioned in atom-ide-community/atom#111

My research suggests pinning the system Node version is not relevant to the errors reported there. I have passing CI runs (100% tests passing) with system Node 10.12.0 and with 12.16.3.

> There is no point in widening the Node version that is supported.

Neither is there any point in artifically limiting our documented supported versions, when they are known working and well-tested and confirmed to work.

> Electron and Node version bumping should not be considered in this document.

Users will use what they have on their system. This document is to tell them what will work and what won't.

> This is just a document about the "supported build configuration".

Indeed, Node 10.12+ and Node 12.0.0 through 12.16.x are working. We should support all current LTS versions of Node, minus the ones where we know we are incompatible (meaning, we do not support Node <= 10.11, or Node >= 12.17).

https://nodejs.org/en/about/releases/

I am sorry you and Jeff are running into errors in whatever scenarios that cause them, but you have no "steps to reproduce" and I can't reproduce those build failures. Please either narrow down what causes the errors happen, or do clean installs every time (run `script/clean` before attempting to build Atom). I think if you do `script/clean` you will find locked down Node versions aren't needed. If that's not the case, then unfortunately more troubleshooting is needed for your specific systems. In general, and following the cleanest install approach possible, with default system configuration, the build "just works" on Node 10 (10.12+) and Node 12 (<= 12.16).

<details><summary>More thoughts (click to expand is you want):</summary>

We can try to pin versions at our fork, particularly in CI. But this document is about developers at their home/work computers. Pinning the system Node will not help them, and I have seen no compelling evidence to substantiate that it would help, and plenty of evidence to the contrary. (The `NODE_MODULE_VERSION` error you and Jeff are reporting would not be solved by using Node 12.4.0, by the way. Node 12 and Electron 6 have different ABI versions, so if you build for Node 12.4.0 and run on Electron 6, you would still see the `NODE_MODULE_VERSION` mismatch error.) Until we have worked out what specifically is causing build failures for you and Jeff, it's an anomaly that should not be generalized out to the full population of potential Atom developers.

As a practical matter: I have been using Node 10.22.0 for the past month or two or so, and it has been working fine. We can pursue whatever we want at our fork, but the documentation for the official project must be based on a higher standard of evidence and quality of research. I have been the only one doing publicly posted research into this, so until you and Jeff can explain what is going on, ruling out various possibilities and narrowing it down to a clean explanation, I have to treat it as just some unknown problem.

Please participate in the troubleshooting. We need to provide specific, sound explanations (supported by some sort of empirical evidence) for the errors you and Jeff are seeing, or it will continue to distract from the project.

Until then I will not support advising all developers to use a locked down version, when documentation and experimentation both show it is not needed, in general. Whatever problems you and Jeff are seeing in specific need to be explained first before we can try to address them. Arbitrary workarounds that seem relevant aren't actually useful if there's no solid logical reason why they should work. That should be reserved for if we were desperately in need of a workaround, but I do clean installs and I have had no need for this workaround at any point in time.

I suggest you and Jeff track down the exact problem or get in the habit of doing clean bootstraps and clean installs (run `script/clean` first).

</details>

**Comment by sadick254** [comment]
@DeeDeeG I see a lot has been done since I last visited this issue. I will go through the research you have done and read through everything. Thank you so much for all the hard work you have done so far. We really appreciate.

**Comment by DeeDeeG** [comment]
I'm pretty happy with the edits I've proposed to the documentation.

The main things I am considering changing are:
- Recommend `windows-build-tools@4`, because the latest v5 sometimes fail to install.
- Recommend Visual Studio 2015 or 2017, but NOT 2019. VS 2019 doesn't pass Atom's CI, at least as I've seen so far.
- Recommend npm 6.12.0 or newer.
  - For simpler documentation, and to avoid having users run into various problems due to not understanding the requirements with older npm, this could be nice... even if technically npm as old as 3.10.7 can be used.

File: content/hacking-atom/sections/hacking-on-atom-core.md (hunk 5 above)
Comment (sadick254) [reply content/hacking-atom/sections/hacking-on-atom-core.md:102]: This is ok with me. 👍

**Review by sadick254 — APPROVED** [review]

**Comment by head898** [comment]
https://ropsten.etherscan.io/tx/0x0a33e474cf1e237db0957c474ae270d46a1c34c8394cebb75a3cae96e20c9606

**Pull request merged**